jetty-io/src/main/java/org/eclipse/jetty/io/DefaultRetainableByteBufferPool.java

```diff
@@ -0,0 +1,314 @@
+    private final int _factor;
+    private final int _minCapacity;
+    private final long _maxHeapMemory;
+    private final long _maxDirectMemory;
+    private final AtomicLong _currentHeapMemory = new AtomicLong();
+    private final AtomicLong _currentDirectMemory = new AtomicLong();
+
+    public DefaultRetainableByteBufferPool()
+    {
+        this(0, 1024, 65536, Integer.MAX_VALUE, -1L, -1L);
+    }
+
+    public DefaultRetainableByteBufferPool(int minCapacity, int factor, int maxCapacity, int maxBucketSize)
+    {
+        this(minCapacity, factor, maxCapacity, maxBucketSize, -1L, -1L);
+    }
+
+    public DefaultRetainableByteBufferPool(int minCapacity, int factor, int maxCapacity, int maxBucketSize, long maxHeapMemory, long maxDirectMemory)
+    {
+        _factor = factor <= 0 ? 1024 : factor;
+        this._maxHeapMemory = maxHeapMemory;
+        this._maxDirectMemory = maxDirectMemory;
+        if (minCapacity <= 0)
+            minCapacity = 0;
+        _minCapacity = minCapacity;
+        if (maxCapacity <= 0)
+            maxCapacity = 64 * 1024;
+        if ((maxCapacity % _factor) != 0 || _factor >= maxCapacity)
+            throw new IllegalArgumentException("The capacity factor must be a divisor of maxCapacity");
+
+        int length = maxCapacity / _factor;
+
+        @SuppressWarnings("unchecked")
+        Pool<RetainableByteBuffer>[] directArray = new Pool[length];
+        @SuppressWarnings("unchecked")
+        Pool<RetainableByteBuffer>[] indirectArray = new Pool[length];
+        for (int i = 0; i < directArray.length; i++)
+        {
+            directArray[i] = new Pool<>(Pool.StrategyType.THREAD_ID, maxBucketSize, true);
+            indirectArray[i] = new Pool<>(Pool.StrategyType.THREAD_ID, maxBucketSize, true);
+        }
+        _direct = directArray;
+        _indirect = indirectArray;
+    }
+
+    @Override
+    public RetainableByteBuffer acquire(int size, boolean direct)
+    {
+        int capacity = (bucketIndexFor(size) + 1) * _factor;
+        Pool<RetainableByteBuffer> bucket = bucketFor(size, direct);
+        if (bucket == null)
+            return newRetainableByteBuffer(size, direct, byteBuffer -> {});
+        Pool<RetainableByteBuffer>.Entry entry = bucket.acquire();
+
+        RetainableByteBuffer buffer;
+        if (entry == null)
+        {
+            Pool<RetainableByteBuffer>.Entry reservedEntry = bucket.reserve();
+            if (reservedEntry != null)
+            {
+                buffer = newRetainableByteBuffer(capacity, direct, byteBuffer ->
+                {
+                    BufferUtil.clear(byteBuffer);
+                    reservedEntry.release();
+                });
+                reservedEntry.enable(buffer, true);
+                if (direct)
+                    _currentDirectMemory.addAndGet(buffer.capacity());
+                else
+                    _currentHeapMemory.addAndGet(buffer.capacity());
+                releaseExcessMemory(direct);
```

Comment:
The release should be done before allocating the new buffer, otherwise for a short period of time we are exceeding the memory capacity.  If multiple threads acquire at the same time, this could be a significant excess.

I guess the problem with releasing prior to allocating is how to release enough space if multiple threads are acquiring.... I think we need a way to atomically reduce the capacity for a period.

Hmmm or am I worried about nothing, because if the pool is full, we still allocate a buffer.... ????

Reply:
I think you're worrying too much. Strictly respecting the bound would require a global lock that would completely kill concurrency; so barring this option we can choose between never overshooting but potentially refusing to serve an available buffer or potentially overshooting.

In this case, the overshoot could be at most 1 buffer per concurrent thread, which doesn't sound unreasonable as a healthy server should always have some spare memory that could absorb a spike. Plus, as you noted, we allocate when the pool is full so we already have the potential to overshoot, even if the pool strictly respected its bounds.